It should not be possible to make a WTF::Function or WTF::CompletionHandler from an Objective-C block in Objective-C++

rr-codes opened the pull request:
#### fc7a43367612f0e0aed65c9e7ceb9539df13b800
<pre>
It should not be possible to make a WTF::Function or WTF::CompletionHandler from an Objective-C block in Objective-C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=300962">https://bugs.webkit.org/show_bug.cgi?id=300962</a>
<a href="https://rdar.apple.com/162838519">rdar://162838519</a>

Reviewed by Abrar Rahman Protyasha.

It&apos;s currently possible to implicitly convert an Objective-C block to a WTF::Function or WTF::CompletionHandler. This
behavior leads to some inconsistencies; while the vast, vast majority of uses use lambdas converted to these types, there
are a handful that use blocks, for no specific reason.

Address these inconsistencies by changing Function.h and CompletionHandler.h to have a new overloaded initializer that
accepts a block, and then delete that initializer. The result of this is that if a client tries to create one of these
types from a block, it will choose this intiializer during overload resolution, and then fail to compile due to the fact
that it is deleted.

This only applies to Objective-C++, since Swift cannot create C++ lambdas itself, so there is no room for ambiguity or inconsistency.

* Source/JavaScriptCore/API/tests/testapi.mm:
(parallelPromiseResolveTest):
* Source/WebCore/platform/ios/wak/WebCoreThread.mm:
(WebThreadRunOnMainThread):
* Source/WebCore/platform/network/cocoa/WebCoreNSURLSession.mm:
(-[WebCoreNSURLSession resetWithCompletionHandler:]):
(-[WebCoreNSURLSession flushWithCompletionHandler:]):
(-[WebCoreNSURLSession getTasksWithCompletionHandler:]):
(-[WebCoreNSURLSession getAllTasksWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView closeAllMediaPresentationsWithCompletionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitialize):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PasteMixedContent.mm:
(TestWebKitAPI::TEST(PasteMixedContent, PasteOneOrMoreURLs)):

- Fix inconsistent uses of blocks as WTF::Function.

* Source/WTF/wtf/CompletionHandler.h:
(WTF::CompletionHandler&lt;Out):
* Source/WTF/wtf/Function.h:
(WTF::Detail::CallableWrapperAllocatorBase::~CallableWrapperAllocatorBase):
(WTF::Function&lt;Out):

- Make it impossible to create these types from blocks in Objective-C++
- Introduce a new base type so that linkage from Swift can work properly

* Source/WTF/wtf/text/StringImpl.h:

- Add missing include

* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPIBundle.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPILibrary.xcconfig:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/SwiftCxxInteropTestbed.cpp: Copied from Tools/TestWebKitAPI/Tests/WebKit Swift/
stWebKitAPIBundle-Bridging-Header.h.
(SwiftCxxInteropTestbed::callIntBoolFunction):
(SwiftCxxInteropTestbed::callIntCompletionHandler):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/SwiftCxxInteropTestbed.h: Copied from Tools/TestWebKitAPI/Tests/WebKit Swift/
stWebKitAPIBundle-Bridging-Header.h.
* Tools/TestWebKitAPI/Tests/WTF/cocoa/SwiftCxxInteropTests.swift: Copied from Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift.
(SwiftCxxInteropTests.wtfFunctionCanBeInvokedFromSwift):
(SwiftCxxInteropTests.wtfCompletionHandlerCanBeInvokedFromSwift):

- Test interop between Swift and WTF C++.

* Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift:
* Tools/TestWebKitAPI/Tests/WebKit Swift/TestWebKitAPIBundle-Bridging-Header.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SmartListsSupport.swift:
(SmartListsTestSelectionConfiguration.caretSelection(withPath:offset:)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:

- Minor clean-up

Canonical link: <a href="https://commits.webkit.org/301813@main">https://commits.webkit.org/301813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c94c1c7c19f8cd14dbc049678124f46ffca7a776

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127041 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134043 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78603 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7482818f-0bf6-43f6-9afe-2e57619a5024) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128912 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55203 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96670 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64689 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d33dbba5-4edd-434c-8837-0882816289a9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129989 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37851 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113683 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77180 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a2d690be-b71d-47f5-95bb-05899330c308) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36718 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31825 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77435 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119078 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107705 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32167 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136569 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125504 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53696 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105185 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54200 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110042 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104876 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26762 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50403 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28780 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51215 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53628 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59501 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158541 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52866 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39655 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56199 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54624 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1300,7 +1309,10 @@ - (void)closeAllMediaPresentations:(void (^)(void))completionHandler
 {
     THROW_IF_SUSPENDED;
-    auto callbackAggregator = CallbackAggregator::create(WTFMove(completionHandler ?: ^{ }));
+    auto callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler)] {
```
I think you can make less changes on this one?
```suggestion
    auto callbackAggregator = CallbackAggregator::create(WTFMove(completionHandler ?: [] { }));
```

rr-codes replied:
I don’t think so, since `completionHandler` would still be a block

rr-codes committed:
It should not be possible to make a WTF::Function or WTF::CompletionHandler from an Objective-C block in Objective-C++
https://bugs.webkit.org/show_bug.cgi?id=300962
rdar://162838519

Reviewed by Abrar Rahman Protyasha.

It's currently possible to implicitly convert an Objective-C block to a WTF::Function or WTF::CompletionHandler. This
behavior leads to some inconsistencies; while the vast, vast majority of uses use lambdas converted to these types, there
are a handful that use blocks, for no specific reason.

Address these inconsistencies by changing Function.h and CompletionHandler.h to have a new overloaded initializer that
accepts a block, and then delete that initializer. The result of this is that if a client tries to create one of these
types from a block, it will choose this intiializer during overload resolution, and then fail to compile due to the fact
that it is deleted.

This only applies to Objective-C++, since Swift cannot create C++ lambdas itself, so there is no room for ambiguity or inconsistency.

* Source/JavaScriptCore/API/tests/testapi.mm:
(parallelPromiseResolveTest):
* Source/WebCore/platform/ios/wak/WebCoreThread.mm:
(WebThreadRunOnMainThread):
* Source/WebCore/platform/network/cocoa/WebCoreNSURLSession.mm:
(-[WebCoreNSURLSession resetWithCompletionHandler:]):
(-[WebCoreNSURLSession flushWithCompletionHandler:]):
(-[WebCoreNSURLSession getTasksWithCompletionHandler:]):
(-[WebCoreNSURLSession getAllTasksWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView closeAllMediaPresentationsWithCompletionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitialize):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PasteMixedContent.mm:
(TestWebKitAPI::TEST(PasteMixedContent, PasteOneOrMoreURLs)):

- Fix inconsistent uses of blocks as WTF::Function.

* Source/WTF/wtf/CompletionHandler.h:
(WTF::CompletionHandler<Out):
* Source/WTF/wtf/Function.h:
(WTF::Detail::CallableWrapperAllocatorBase::~CallableWrapperAllocatorBase):
(WTF::Function<Out):

- Make it impossible to create these types from blocks in Objective-C++
- Introduce a new base type so that linkage from Swift can work properly

* Source/WTF/wtf/text/StringImpl.h:

- Add missing include

* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPIBundle.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPILibrary.xcconfig:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/SwiftCxxInteropTestbed.cpp: Copied from Tools/TestWebKitAPI/Tests/WebKit Swift/
stWebKitAPIBundle-Bridging-Header.h.
(SwiftCxxInteropTestbed::callIntBoolFunction):
(SwiftCxxInteropTestbed::callIntCompletionHandler):
* Tools/TestWebKitAPI/Tests/WTF/cocoa/SwiftCxxInteropTestbed.h: Copied from Tools/TestWebKitAPI/Tests/WebKit Swift/
stWebKitAPIBundle-Bridging-Header.h.
* Tools/TestWebKitAPI/Tests/WTF/cocoa/SwiftCxxInteropTests.swift: Copied from Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift.
(SwiftCxxInteropTests.wtfFunctionCanBeInvokedFromSwift):
(SwiftCxxInteropTests.wtfCompletionHandlerCanBeInvokedFromSwift):

- Test interop between Swift and WTF C++.

* Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift:
* Tools/TestWebKitAPI/Tests/WebKit Swift/TestWebKitAPIBundle-Bridging-Header.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SmartListsSupport.swift:
(SmartListsTestSelectionConfiguration.caretSelection(withPath:offset:)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:

- Minor clean-up

Canonical link: https://commits.webkit.org/301813@main

aprotyas commented on Source/WTF/wtf/Function.h:
```diff
@@ -76,6 +82,19 @@ class Function<Out(In...)> {
     Function(FunctionType f)
         : m_callableWrapper(makeUnique<Detail::CallableWrapper<FunctionType, Out, In...>>(std::forward<FunctionType>(f))) { }
 
+#if defined(__APPLE__)
```
turns out there is a better proxy! `__BLOCKS__`
